### PR TITLE
opengl: use requiredKernelConfig

### PR DESCRIPTION
**Pull request by jhvst**
###### Description of changes

*Expectation*: When building a kernel using `manualConfig`, OpenGL can be built when the option [`IA32_EMULATION`](https://www.kernelconfig.io/CONFIG_IA32_EMULATION?q=IA32_EMULATION&kernelversion=5.18.18&arch=x86) is yes.

*What actually happens*: OpenGL insists that `IA32_EMULATION` is no with an error message `Option driSupport32Bit requires a kernel that supports 32bit emulation`.

*Description of the PR*: Update the OpenGL package to use `requiredKernelConfig` to check 32bit emulation support. Most packages, e.g. [`zram`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/zram.nix#L122) and [`systemd`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/systemd.nix#L566) use this (seemingly) newer approach, and I imagine that OpenGL should be updated accordingly.

*Related links*: Discussion previously on a Russian Linux forum: [NixOS: dri 32 бит на своем ядре](https://www.linux.org.ru/forum/general/16714583)

Things I haven't checked:
* if there is any problem with other platforms than Linux
* if this conflicts with some other build process in which the kernel is not built (i.e., is `kernelConfig` always available)

Note: `pkgs.stdenv.isx86_64` check is redundant when checking for the `IA32_EMULATION` option, as the option depends on [`CONFIG_X86_64`](https://www.kernelconfig.io/CONFIG_X86_64?kernelversion=5.18.18) being true.

Also, please do squash the commits for me if you end up merging.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/hardware/opengl.nix [hunk 1]
```diff
@@ -125,14 +125,10 @@ in
...
-        message = "Option driSupport32Bit only makes sense on a 64-bit system.";
-      }
-      { assertion = cfg.driSupport32Bit -> (config.boot.kernelPackages.kernel.features.ia32Emulation or false);
-        message = "Option driSupport32Bit requires a kernel that supports 32bit emulation";
-      }
-    ];
+    system.requiredKernelConfig =
+      if cfg.driSupport32Bit
+      then map config.lib.kernelConfig.isYes [ "IA32_EMULATION" ]
+      else null;
```
Comment (rnhmjoj) [inline nixos/modules/hardware/opengl.nix:131]: Did you test this? The option accepts a list, `null` shouldn't work.
Anyway, I suggest

```suggestion
    system.requiredKernelConfig =
      lib.optional cfg.driSupport32Bit 
      (config.lib.kernelConfig.isYes "IA32_EMULATION")
```

Comment (jhvst) [reply nixos/modules/hardware/opengl.nix]: It did work (surprisingly), but I wasn't very proud of it -- I'm new to Nix. This seems much better.

**Comment by rnhmjoj** [comment]
> if there is any problem with other platforms than Linux

NixOS only supports the Linux kernel, so no need to check for anything else.
I wonder what it does on other architectures like aarch64, which is supported.
But still, that option probably never intended to support anything other than x86_64.

> if this conflicts with some other build process in which the kernel is not built (i.e., is kernelConfig always available)

Good question: I looked at the code and while `requiredKernelConfig` is always defined `lib.kernelConfig` is not available in a container (`config.boot.isCOntainer`). However we don't have to worry because about that because the checks are also skipped in that case.

**Comment by jhvst** [comment]
Gotcha, in that case, is there anything else to do here? That is, while I'm glad that this change seems to be directed upstream thus fixing *my* problem, it does leave me wondering whether the whole `kernel.features.ia32Emulation` construct is redundant. I'd guess no (and I think you know for better), which leaves me thinking there *might* exist a reason why the `requiredKernelConfig` was not introduced here in the first place.

**Comment by rnhmjoj** [comment]
> Gotcha, in that case, is there anything else to do here?

No, looks good to me. I just want to test it properly

> leaves me thinking there _might_ exist a reason why the `requiredKernelConfig` was not introduced here in the first place.

The kernel is not my area of expertice, but I think `kernel.features` is a mechanism to toggle some predefined sets of kernel build options (see `pkgs/os-specific/linux/kernel/common-config.nix`) and signal that said options are on/off (via a `passthru`) to NixOS modules. Kernels built with `manualConfig` lacks `features` and that's the reason why NixOS reported incorrectly that you lacked x86 emulation.

You can see some of these `features` in `pkgs/os-specific/linux/kernel/generic.nix`, however outside of `netfilterRPFilter` they are all unused. My guess is that this mechanism (the signalling part) is obsolete and should be removed since `requiredKernelConfig` can handle both common/manual config.

By the way this isn't the first time this happened: 534998e005c2bc7d24edb237052151e38e1946eb.

File: nixos/modules/hardware/opengl.nix [hunk 2]
```diff
@@ -125,14 +125,9 @@ in
...
-      { assertion = cfg.driSupport32Bit -> pkgs.stdenv.isx86_64;
-        message = "Option driSupport32Bit only makes sense on a 64-bit system.";
-      }
-      { assertion = cfg.driSupport32Bit -> (config.boot.kernelPackages.kernel.features.ia32Emulation or false);
-        message = "Option driSupport32Bit requires a kernel that supports 32bit emulation";
-      }
-    ];
+    system.requiredKernelConfig =
+      lib.optional cfg.driSupport32Bit
+      (config.lib.kernelConfig.isYes "IA32_EMULATION");
```
Comment (SuperSandro2000) [inline nixos/modules/hardware/opengl.nix:130]:
```suggestion
    system.requiredKernelConfig = lib.optional cfg.driSupport32Bit (config.lib.kernelConfig.isYes "IA32_EMULATION");
```

**Comment by rnhmjoj** [comment]
So, I managed to test this change with this configuration:

```
{ pkgs, lib, ... }:

{
  fileSystems."/".label = "root";
  boot.loader.grub.enable = false;

  hardware.opengl.enable = true;
  hardware.opengl.driSupport = true;
  hardware.opengl.driSupport32Bit = true;

  boot.kernelPackages = pkgs.linuxPackagesFor
    (pkgs.linuxKernel.manualConfig
      { inherit (pkgs) stdenv lib;
        inherit (pkgs.linuxPackages_latest.kernel) src version;
        configfile = /home/rnhmjoj/config;
        allowImportFromDerivation = true;
      });
}
```
It works in the sense that you get the error if `IA32_EMULATION` is off and no error otherwise.
However there are several catches:

1. `requiredKernelConfig` ignores any assertio if the kernel has the `features` attribute:
    ```
    # nixpkgs kernels are assumed to have all required features
    assertions = if config.boot.kernelPackages.kernel ? features then [] else ...
    ```

    So, if you do something like this:
    ```
      hardware.opengl.driSupport32Bit = true;
      boot.kernelPatches = lib.singleton
        { name = "no-32bit-emulation";
          patch = null;
          extraConfig = "IA32_EMULATION n";
        };
    ```
    you wont get the error.

2. `requiredKernelConfig` works by inspecting the attribute `kernel.config`, which
    is only generated if you set `allowImportFromDerivation = true`. If you don't
    do this you won't get error.

3. The error is pretty terrible compared to the current one: 

   ```
   Failed assertions:
   - CONFIG_IA32_EMULATION is not yes!
   ``` 

   There's currently no way to provide a message to explain what is the source
   of the error and why that is a problem.
   We could avoid this by manually writing the assertion (using `kernel.config.isYes`),
   but I'd rather fix this in general. An easy way would be to define a new function
   (name to be defined) like
   ```
   requiredKernelConfig = [
      (config.lib.kernelConfig.isYesForReason
        { option = "IA32_EMULATION"; 
          reason = "Option hardware.opengl.driSupport32Bit requires a kernel that supports 32bit emulation";
        })
   ];
   ```

**Comment by jhvst** [comment]
Thanks for the testing @rnhmjoj. You hit some deficiencies I found when building a minimal custom kernel. Long story short, it's easy to miss required features as few packages know their kernel requirements, and while there exists a range of options how to apply kernel configuration options, the ordering when such changes are applied and become visible is confusing. At the current state of affairs, builds which use a custom kernel could be accidentally broken (with your patch example being, I think, a perfectly reasonable source of configuration conflict, and which may quite a lot e.g. on nixos-mobile configurations).

Few comments on your points, given my short experience with kernel configuration:

1. The `features` option seems confusing, because it seems to introduce extra configuration into the kernel which is not derivable from the applied NixOS packages. That is, I think in a perfect world each of the NixOS packages know what kernel options are required, and these requirements cascade into feature sets automatically, thus avoiding the need for the `features` to exist. Further, given the existence of `kernelPatches`, there seems to be a way to apply arbitrary kernel configuration for those who do need it. Though, changing `features` is much beyond this change, but I will open a separate issue about this once I find the time, as I ended up finding a lot of implicit kernel configuration requirements hence think there's room for improvement.

2. Aside what I said above, I don't think a custom kernel can be compiled without `allowImportFromDerivation = true`, thus this could be considered fine. Why is this required, I don't know. 

3. I agree this would be helpful, but even more helpful would be to have a flag with `manualConfig` to apply the required changes automatically. Nevertheless, it'd be nice to know the source package which produced the error, because when you get 30 of these, it'd be nice to know which package caused the most requirements. But also, have an optional comment field explaining why something must be the way it is. E.g., for `zram` to work it must be a module (instead of a `yes`), because modularization allows the parameters to be re-set during boot. Similar situations exist for the opposite case. I think the reasoning for these cases would be great to come in the source code, and not only as wiki page knowledge at Gentoo's.

So, aside of the style change requested above, should all the three set aside for now? I can open separate issues for each, referencing this PR, and maybe people with more insight into the implementation can comment on them.

**Comment by rnhmjoj** [comment]
1. I agree, the `features` attribute shouldn't exist, ideally. The problem is that I don't think it's possible to generate the `config` attrset for the standard NixOS kernel without doing an [IFD](https://nixos.wiki/wiki/Import_From_Derivation), which regarded as a bad thing.

2. I see. I think it's needed, again, because the `config` attset is generated by parsing the kernel config file.
To do this this you need IFD: among other thing, this is not allowed in the restricted evaluation mode used by hydra, the NixOS build farm, so you wouldn't be able to build and distribute the kernel.

3. Uhm, I would insist on improving the error message. It's true that this PR fixes an hard failure when using a `manualConfig` kernel, but it degrades the experience for users of the normal kernels: it's replacing the current error message with a very abstruse one.

**Comment by jhvst** [comment]
Updating on this as I'm ran into it again. Seemingly, the error message is blocking this PR.

The printing happens here: https://github.com/NixOS/nixpkgs/blob/9477fa1e44c16804b8c1743d25264f9e3b26c219/nixos/modules/system/activation/top-level.nix#L114

The pattern matching is here: https://github.com/NixOS/nixpkgs/blob/75025ad5e1c991f13c47dd283d69ae852b38e2a2/pkgs/os-specific/linux/kernel/manual-config.nix#L64

The implementation are here: https://github.com/NixOS/nixpkgs/blob/9477fa1e44c16804b8c1743d25264f9e3b26c219/nixos/modules/system/boot/kernel.nix#L297

I suggest to extend `isYes` with an optional argument for finer description.

```
 isYes = option (optional reason): {
            assertion = config: config.isYes option;
            message = "CONFIG_${option} is not yes!" ++ reason ? " ${reason}" : "";
            configLine = "CONFIG_${option}=y";
};
```

I don't know much about Nix standard library so no idea how to write this in a good manner, but you get the idea. @rnhmjoj do you think this is a good abstraction to extend? I think anything else is the wrong place to implement this change. Allowing other assertions to also include more data seems like useful outside of this PR.

**Comment by jhvst** [comment]
I have moved on to use different kind of kernel configuration, i.e., the one in which I supply `kernelPatches`: https://github.com/jhvst/nix-config/blob/0bbeb07fff704f711abfea15dd49288242bc76fc/hosts/muro/default.nix#L64

Reason for this is that manual kernel configuration caused more regression that they did good.

For this reason, this PR is hugely stale now, for which I am closing this issue.

**Pull request closed without merging**